File: bedrock/src/games/extras/logic/wheel.ts
/**
 * Wheel of Fortune (GAME_DESIGN §11.2, Appendix B). PURE.
 * 54 segments in a fixed order (index 0 at the pointer, clockwise); a spin lands uniformly on
 * one index and pays floor(stake × multiplier). The Creeper segment pays 0 and triggers the
 * chaos `mob_wave` event.
 */
import { type Rng } from '../../../core/logic/rng';
import { floorPay } from './payout';

export type WheelCode = 'B' | 'C' | 'H' | 'M' | 'D' | 'T' | 'E' | 'X';
export const WHEEL_CODES: readonly WheelCode[] = ['B', 'C', 'H', 'M', 'D', 'T', 'E', 'X'];

/** Appendix B, index 0…53 clockwise from the pointer at rest. */
export const DEFAULT_SEGMENTS: readonly WheelCode[] = (
  'X B M B D B M B H B D B M B T B M B D B H B M B D B E ' + 'B M B D B H B M B T B M B D B H B M B T B C M H D M B'
).split(' ') as WheelCode[];

export const DEFAULT_MULTIPLIERS: Readonly<Record<WheelCode, number>> = { B: 0, C: 0, H: 0.5, M: 1, D: 2, T: 3, E: 5, X: 10 };

/** Lang key suffix per code: `gui.burmaldaholic.extras.wheel.segment.<name>`. */
export const SEGMENT_NAMES: Readonly<Record<WheelCode, string>> = {
  B: 'bust',
  C: 'creeper',
  H: 'half',
  M: 'money_back',
  D: 'double',
  T: 'triple',
  E: 'emerald',
  X: 'diamond',
};
export const segmentKey = (c: WheelCode): string => `gui.burmaldaholic.extras.wheel.segment.${SEGMENT_NAMES[c]}`;

/** Color per segment (UI.md: loss §c, push §7, win §a, jackpot §6). */
export const SEGMENT_COLORS: Readonly<Record<WheelCode, string>> = { B: '§c', C: '§2', H: '§c', M: '§7', D: '§a', T: '§a', E: '§6', X: '§b' };

export interface WheelSetup {
  segments: readonly WheelCode[];
  multipliers: Readonly<Record<WheelCode, number>>;
}

export const DEFAULT_WHEEL: WheelSetup = { segments: DEFAULT_SEGMENTS, multipliers: DEFAULT_MULTIPLIERS };

const isCode = (x: unknown): x is WheelCode => typeof x === 'string' && (WHEEL_CODES as readonly string[]).includes(x);

/**
 * Build a wheel from config values (`extras.wheel.segments`, `extras.wheel.multipliers`).
 * Unknown codes are dropped; a wheel with fewer than 2 valid segments falls back to Appendix B.
 * Missing multipliers use the defaults; negative / non-finite ones become 0.
 */
export function wheelFromConfig(segments: unknown, multipliers: unknown): WheelSetup {
  const segs = Array.isArray(segments) ? segments.filter(isCode) : [];
  const mult: Record<WheelCode, number> = { ...DEFAULT_MULTIPLIERS };
  if (multipliers && typeof multipliers === 'object') {
    for (const [k, v] of Object.entries(multipliers as Record<string, unknown>)) {
      if (isCode(k)) mult[k] = typeof v === 'number' && Number.isFinite(v) && v > 0 ? v : 0;
    }
  }
  return { segments: segs.length >= 2 ? segs : DEFAULT_SEGMENTS, multipliers: mult };
}

export interface WheelSpin {
  index: number;
  code: WheelCode;
  multiplier: number;
}

export function spinWheel(rng: Rng, w: WheelSetup = DEFAULT_WHEEL): WheelSpin {
  const index = Math.floor(rng.next() * w.segments.length);
  const code = w.segments[index] as WheelCode;
  return { index, code, multiplier: w.multipliers[code] };
}

/** Total return (stake included). */
export const wheelReturn = (stake: number, spin: Pick<WheelSpin, 'multiplier'>): number => floorPay(stake, spin.multiplier);

/** Segment counts in wheel order of first appearance. */
export function segmentCounts(w: WheelSetup = DEFAULT_WHEEL): Map<WheelCode, number> {
  const m = new Map<WheelCode, number>();
  for (const c of w.segments) m.set(c, (m.get(c) ?? 0) + 1);
  return m;
}

/** Legend rows ordered by multiplier (bust first), for the form body. */
export function wheelLegend(w: WheelSetup = DEFAULT_WHEEL): { code: WheelCode; multiplier: number; count: number }[] {
  const counts = segmentCounts(w);
  return WHEEL_CODES.filter((c) => counts.has(c))
    .map((c) => ({ code: c, multiplier: w.multipliers[c], count: counts.get(c) ?? 0 }))
    .sort((a, b) => a.multiplier - b.multiplier || (a.code === 'C' ? 1 : 0) - (b.code === 'C' ? 1 : 0));
}

/** Exact RTP (before flooring): mean multiplier over the segments. */
export function wheelRtp(w: WheelSetup = DEFAULT_WHEEL): number {
  return w.segments.reduce((s, c) => s + w.multipliers[c], 0) / w.segments.length;
}

export const maxWheelMultiplier = (w: WheelSetup = DEFAULT_WHEEL): number => Math.max(...w.segments.map((c) => w.multipliers[c]));

/**
 * Pointer path for the spin animation: `turns` full turns then decelerate onto `target`.
 * Returns the indices to show (last = target) and the tick delay before each frame
 * (growing, so the wheel slows down). Deterministic.
 */
export function spinFrames(target: number, size: number, frames = 24, turns = 1): { index: number; delay: number }[] {
  const n = Math.max(2, size);
  const total = turns * n + (((target % n) + n) % n);
  const out: { index: number; delay: number }[] = [];
  for (let f = 1; f <= frames; f++) {
    // ease-out: position = total × (1 − (1 − f/frames)²)
    const x = f / frames;
    const pos = Math.round(total * (1 - (1 - x) * (1 - x)));
    out.push({ index: pos % n, delay: 1 + Math.floor(5 * x * x) });
  }
  out[out.length - 1] = { index: ((target % n) + n) % n, delay: out[out.length - 1]?.delay ?? 6 };
  return out;
}

File: java/src/main/java/dev/nezo/burmaldaholic/games/extras/logic/Wheel.java
package dev.nezo.burmaldaholic.games.extras.logic;

import dev.nezo.burmaldaholic.core.rng.CasinoRng;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Wheel of Fortune (GAME_DESIGN.md §11.2, appendix B). The segments are fixed in order (index 0 at the
 * pointer, clockwise); a spin lands uniformly on one index and pays {@code floor(stake × multiplier)}.
 * The Creeper segment pays 0 and triggers the chaos {@code mob_wave} event. Pure Java.
 */
public final class Wheel {
	public static final List<String> CODES = List.of("B", "C", "H", "M", "D", "T", "E", "X");
	public static final List<String> DEFAULT_SEGMENTS = List.of(
		"X", "B", "M", "B", "D", "B", "M", "B", "H", "B", "D", "B", "M", "B", "T", "B", "M", "B", "D", "B", "H", "B", "M", "B", "D", "B", "E",
		"B", "M", "B", "D", "B", "H", "B", "M", "B", "T", "B", "M", "B", "D", "B", "H", "B", "M", "B", "T", "B", "C", "M", "H", "D", "M", "B");
	public static final Map<String, Double> DEFAULT_MULTIPLIERS = Map.of("B", 0.0, "C", 0.0, "H", 0.5, "M", 1.0, "D", 2.0, "T", 3.0, "E", 5.0, "X", 10.0);
	public static final String CREEPER = "C";

	private final List<String> segments;
	private final Map<String, Double> multipliers;
	private final List<CasinoRng.Weighted<Integer>> uniform;

	public Wheel(List<String> segments, Map<String, Double> multipliers) {
		Map<String, Double> m = new LinkedHashMap<>(DEFAULT_MULTIPLIERS);
		if (multipliers != null) {
			multipliers.forEach((k, v) -> {
				if (k != null && v != null) {
					m.put(k, Double.isFinite(v) && v > 0 ? v : 0.0);
				}
			});
		}
		List<String> segs = new ArrayList<>();
		if (segments != null) {
			for (String s : segments) {
				if (s != null && m.containsKey(s)) {
					segs.add(s);
				}
			}
		}
		this.segments = List.copyOf(segs.size() >= 2 ? segs : DEFAULT_SEGMENTS);
		this.multipliers = Collections.unmodifiableMap(m);
		List<CasinoRng.Weighted<Integer>> all = new ArrayList<>(this.segments.size());
		for (int i = 0; i < this.segments.size(); i++) {
			all.add(new CasinoRng.Weighted<>(i, 1.0));
		}
		this.uniform = List.copyOf(all);
	}

	public static Wheel standard() {
		return new Wheel(DEFAULT_SEGMENTS, DEFAULT_MULTIPLIERS);
	}

	public List<String> segments() {
		return segments;
	}

	public int size() {
		return segments.size();
	}

	public double multiplier(String code) {
		return multipliers.getOrDefault(code, 0.0);
	}

	public record Spin(int index, String code, double multiplier) {
		public boolean creeper() {
			return CREEPER.equals(code);
		}
	}

	public Spin at(int index) {
		String code = segments.get(Math.floorMod(index, segments.size()));
		return new Spin(Math.floorMod(index, segments.size()), code, multiplier(code));
	}

	/**
	 * One spin: uniform over the segments. Drawn with {@link CasinoRng#weighted} so modifiers (VIP, chaos,
	 * Last Chance) can tilt the share of paying segments ({@code multiplier > 1}); without modifiers it is
	 * exactly uniform.
	 */
	public Spin spin(CasinoRng rng) {
		int index = rng.weighted(uniform, i -> multiplier(segments.get(i)) > 1.0);
		return at(index);
	}

	/** Total return (stake included). */
	public static long totalReturn(long stake, Spin spin) {
		return Payouts.floorPay(stake, spin.multiplier());
	}

	/** Exact RTP before flooring: the mean multiplier. */
	public double rtp() {
		double sum = 0;
		for (String s : segments) {
			sum += multiplier(s);
		}
		return sum / segments.size();
	}

	public double maxMultiplier() {
		double max = 0;
		for (String s : segments) {
			max = Math.max(max, multiplier(s));
		}
		return max;
	}

	public Map<String, Integer> counts() {
		Map<String, Integer> m = new LinkedHashMap<>();
		for (String s : segments) {
			m.merge(s, 1, Integer::sum);
		}
		return m;
	}

	public record LegendRow(String code, double multiplier, int count) {}

	/** Legend rows ordered by multiplier (bust first, creeper after bust). */
	public List<LegendRow> legend() {
		List<LegendRow> rows = new ArrayList<>();
		counts().forEach((code, n) -> rows.add(new LegendRow(code, multiplier(code), n)));
		rows.sort(Comparator.comparingDouble(LegendRow::multiplier).thenComparing(r -> CREEPER.equals(r.code()) ? 1 : 0));
		return rows;
	}

	/** Lang key suffix per code: {@code gui.burmaldaholic.extras.wheel.segment.<name>}. */
	public static String segmentKey(String code) {
		String name = switch (code) {
			case "B" -> "bust";
			case "C" -> "creeper";
			case "H" -> "half";
			case "M" -> "money_back";
			case "D" -> "double";
			case "T" -> "triple";
			case "E" -> "emerald";
			case "X" -> "diamond";
			default -> "bust";
		};
		return "gui.burmaldaholic.extras.wheel.segment." + name;
	}

	/**
	 * Segment name without its multiplier, for lines that print the multiplier next to it (the legend):
	 * {@code segment.emerald} is "Emerald ×5", the legend shows "Emerald — ×5".
	 */
	public static String segmentNameKey(String code) {
		return switch (code) {
			case "E" -> "gui.burmaldaholic.extras.wheel.segment_name.emerald";
			case "X" -> "gui.burmaldaholic.extras.wheel.segment_name.diamond";
			default -> segmentKey(code);
		};
	}

	/** ARGB color per segment code (loss red, push gray, win green, big wins gold/aqua). */
	public static int color(String code) {
		return switch (code) {
			case "B" -> 0xFF8E2A2A;
			case "C" -> 0xFF2E8B2E;
			case "H" -> 0xFFB5563A;
			case "M" -> 0xFF7A7A7A;
			case "D" -> 0xFF3A9E4A;
			case "T" -> 0xFF2F7FC0;
			case "E" -> 0xFF18B45A;
			case "X" -> 0xFF4FD8E0;
			default -> 0xFF444444;
		};
	}
}
